contracts/Pool.sol

```diff
@@ -1,31 +1,292 @@
+        if (endBlock <= block.number) {
+            return sender.startBalance % amtPerBlock;
+        }
+        return
+            sender.startBalance -
+            (block.number - sender.startBlock) *
+            amtPerBlock;
     }
 
-    function topUpSender(address sender) public payable {
-        senders[sender].balance += msg.value;
+    /// @notice Withdraws unsent funds of the sender of the message and sends them to that sender
+    /// @param amount The amount to be withdrawn, must not be higher than available funds
+    function withdraw(uint256 amount) public pauseSending {
+        uint192 startBalance = senders[msg.sender].startBalance;
+        require(
+            amount <= startBalance,
+            "Not enough funds in the sender account"
+        );
+        senders[msg.sender].startBalance = startBalance - uint192(amount);
+        msg.sender.transfer(amount);
+    }
+
+    /// @notice Sets the target amount sent on every block from the sender of the message.
+    /// On every block this amount is rounded down to the closest multiple of the sum of the weights
+    /// of the receivers and split between all sender's receivers proportionally to their weights.
+    /// Each receiver receives then their part from the sender's balance.
+    /// If set to zero, stops sending any funds.
+    /// @param amount The target per-block amount
+    function setAmountPerBlock(uint256 amount) public pauseSending {
+        senders[msg.sender].amtPerBlock = uint192(amount);
+    }
+
+    /// @notice Sets the weight of a receiver of the sender of the message.
+    /// The weight regulates the share of the amount being sent on every block in relation to
+    /// other sender's receivers.
+    /// Setting a non-zero weight for a new receiver, added it to the list of sender's receivers.
+    /// Setting the zero weight for a receiver, removes it from the list of sender's receivers.
+    /// @param receiver The address of the receiver
+    /// @param weight The weight of the receiver
+    function setReceiver(address receiver, uint32 weight) public pauseSending {
+        Sender storage sender = senders[msg.sender];
+        uint32 oldWeight = sender.receiverWeights.setWeight(receiver, weight);
+        sender.weightSum -= oldWeight;
+        sender.weightSum += weight;
+        require(
+            sender.weightSum <= SENDER_WEIGHTS_SUM_MAX,
+            "Too much total receivers weight"
+        );
+        if (weight != 0 && oldWeight == 0) {
+            sender.weightCount++;
+            require(
+                sender.weightSum <= SENDER_WEIGHTS_COUNT_MAX,
+                "Too many receivers"
+            );
+        } else if (weight == 0 && oldWeight != 0) {
+            sender.weightCount--;
+        }
+    }
+
+    /// @notice Pauses all sender's payments on current block allowing safe configuration
+    modifier pauseSending {
+        stopSending();
+        _;
+        startSending();
+    }
```

Comment:
Can you explain why this is needed?

Reply:
It's a modifier, which is used on all sender modifier functions. It assures that all changes applied to sender are applied with a clean state, with no payment happening on that sender and with no risk of messing it up. It may be optimized away for `topUp` to make this common operation cheaper.

Reply:
What you explain sounds reasonable but I'd never expected this code to be ensuring that :exploding_head:

Reply:
Hmm, but why aren't things in a clean state in the first place?

Reply:
@NunoAlexandre This is a modifier, see https://solidity.readthedocs.io/en/v0.6.12/contracts.html#modifiers

Reply:
@cloudhead Because the receivers' deltas are shifted to account for the currently ongoing payment, see `startSending`

Reply:
Looks resolved